doc: net: Add network configuration guide

jukkar opened the pull request:
Add a simple document describing various network related configuration options and how they affect the available resources in the system.

jukkar committed:
doc: net: Add network configuration guide

Add a simple document describing various network related
configuration options and how they affect the available
resources in the system.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on doc/connectivity/networking/net_config_guide.rst:
```diff
@@ -0,0 +1,131 @@
+   below 9, though.
+   Should a retransmission timeout occur, the receive callback is
+   called with -ECONNRESET error code and the context is dereferenced."
+   ":kconfig:option:`CONFIG_NET_TCP_MAX_SEND_WINDOW_SIZE`",
```
I think we should elaborate here on the impact of the window size on the net_buf usage. I. e. that this value might require finetuning (lowering) when it is expected to have multiple active TCP connections in the system, otherwise multiple TCP connections could easily exhaust net_buf pool for the queued data.

rlubos reviewed (commented):
Good starting point. I wonder if it'd make sense to group those settings in separate sections, i.e. one for generic networking configuration, and then separate for sockets, TLS sockets, TCP.

kartben commented on doc/connectivity/networking/net_config_guide.rst:
```diff
@@ -0,0 +1,131 @@
+This document describes how various network configuration options can be
+set according to available resources in the system.
+
+.. csv-table:: network configuration options
```
a ``list-table`` would be easier to author / maintain going forward, and would not necessitate the use of double quotes

jukkar commented on doc/connectivity/networking/net_config_guide.rst:
```diff
@@ -0,0 +1,131 @@
+   ":kconfig:option:`CONFIG_NET_MAX_CONN`",
+   "This option tells how many network connection endpoints are supported.
+   For example each TCP connection requires one connection endpoint. Similarly
+   each listening UDP connection requires one connection endpoint."
```
Yep, I will also add mention that net-shell "net conn" command can be used to see the available connections at runtime.

jukkar commented on doc/connectivity/networking/net_config_guide.rst:
```diff
@@ -0,0 +1,131 @@
+   values are in the 0-31 range.  It's highly recommended to not go
+   below 9, though.
+   Should a retransmission timeout occur, the receive callback is
+   called with -ECONNRESET error code and the context is dereferenced."
```
Nice catch. I copied this text from tcp Kconfig file, will fix it there too.

jukkar committed:
net: tcp: Fix the CONFIG_NET_TCP_RETRY_COUNT help text

The help text was incorrect, we return -ETIMEDOUT instead
of -ECONNRESET when retransmission timeout occurs.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
> I wonder if it'd make sense to group those settings in separate sections, i.e. one for generic networking configuration, and then separate for sockets, TLS sockets, TCP.

Yes, that sounds like a good plan. Lets first get enough right content, and then fine tune the output.

jukkar commented:
The list format is not looking good, https://builds.zephyrproject.io/zephyr/pr/67239/docs/connectivity/networking/net_config_guide.html
I will try to figure out better format.

kartben commented:
> The list format is not looking good, https://builds.zephyrproject.io/zephyr/pr/67239/docs/connectivity/networking/net_config_guide.html I will try to figure out better format.

Try "term list", perhaps https://docs.zephyrproject.org/latest/develop/optimizations/footprint.html#stack-sizes

jukkar commented:
> Try "term list", perhaps https://docs.zephyrproject.org/latest/develop/optimizations/footprint.html#stack-sizes

Indeed, that looks reasonable.

rlubos commented on doc/connectivity/networking/net_config_guide.rst:
```diff
@@ -0,0 +1,267 @@
+  This option is used by socketpair() function. It sets the size of the
+  internal intermediate buffer, in bytes. This sets the limit how large
+  messages can be passed between two socketpair endpoints.
+
```
Maybe we should mention `CONFIG_POSIX_MAX_FDS` here as well, as this parameter also limits the number of sockets that can be created in the system.

rlubos reviewed (dismissed):
Very nice! Left one suggestion, but it looks very good already as it is.

kartben commented on doc/connectivity/networking/net_config_guide.rst:
```diff
@@ -0,0 +1,272 @@
+  The following formula can be used to determine the time (in ms)
+  that a segment will be be buffered awaiting retransmission::
+
+     n=NET_TCP_RETRY_COUNT
+     Sum((1<<n) * NET_TCP_INIT_RETRANSMISSION_TIMEOUT)
+     n=0
```
non blocking but I recommend using `math` directive.
See ex. https://docs.zephyrproject.org/latest/boards/posix/native_sim/doc/index.html#how-simulated-time-and-real-time-relate-to-each-other https://github.com/zephyrproject-rtos/zephyr/blob/5c34726ab11de79d8c5a456153d81d40fa3124eb/boards/posix/native_sim/doc/index.rst#L273-L290

jukkar replied:
I can investigate if it is easy to add sigma symbol here, or just use the current text but have the math directive.

kartben replied:
something like this?

```math
\sum_{n=0}^{\text{NET\_TCP\_RETRY\_COUNT}} (1 \ll n) \times \text{NET\_TCP\_INIT\_RETRANSMISSION\_TIMEOUT}
```

```
\sum_{n=0}^{\text{NET\_TCP\_RETRY\_COUNT}} (1 \ll n) \times \text{NET\_TCP\_INIT\_RETRANSMISSION\_TIMEOUT}
```

kartben commented on doc/connectivity/networking/net_config_guide.rst:
```diff
@@ -0,0 +1,272 @@
+  below 9, though.
+
+  Should a retransmission timeout occur, the receive callback is
+  called with *-ETIMEDOUT* error code and the context is dereferenced.
```
```suggestion
  called with ``-ETIMEDOUT`` error code and the context is dereferenced.
```

or 

```suggestion
  called with :code:`-ETIMEDOUT` error code and the context is dereferenced.
```

would be better

kartben reviewed (dismissed):
Only non-blocking comments from me -- thanks!

One additional note: I am assuming you have listed all the Kconfig options in a deliberate order (in order of importance?) ; but if that is not the case, I wonder if it would make sense to list them in alphabetical order in each section as it would ensure everything is (and remains, going forward) grouped somewhat logically.

jukkar commented:
> I am assuming you have listed all the Kconfig options in a deliberate order (in order of importance?)

Yes, that is correct. I would like to keep the ordering as it makes more sense here.

jukkar commented:
Updated according to comments.

kartben commented on doc/connectivity/networking/net_config_guide.rst:
```diff
@@ -0,0 +1,273 @@
+:kconfig:option:`CONFIG_NET_TCP_RETRY_COUNT`
+  Maximum number of TCP segment retransmissions.
+  The following formula can be used to determine the time (in ms)
+  that a segment will be be buffered awaiting retransmission::
+
+  .. math::
+
+     \sum_{n=0}^{n=NET\_TCP\_RETRY\_COUNT} \bigg(1<<n\bigg)\times
+     NET\_TCP\_INIT\_RETRANSMISSION\_TIMEOUT
```
I seem to remember an issue with PDF rendering if literals containing underscores where not using `\mathtt` (the formulas are rendered as "LaTeX equations" which IIRC means some constructs are treated differently).

```suggestion
  that a segment will be be buffered awaiting retransmission:

  .. math::

     \sum_{n=0}^{\mathtt{NET\_TCP\_RETRY\_COUNT}} \bigg(1<<n\bigg)\times
     \mathtt{NET\_TCP\_INIT\_RETRANSMISSION\_TIMEOUT}
```

kartben commented on doc/connectivity/networking/net_config_guide.rst:
```diff
@@ -0,0 +1,273 @@
+:kconfig:option:`CONFIG_NET_TCP_RETRY_COUNT`
+  Maximum number of TCP segment retransmissions.
+  The following formula can be used to determine the time (in ms)
+  that a segment will be be buffered awaiting retransmission::
```
You probably missed it in the previous comments (my bad for not having making it clearer) but the extra ":" needs to go otherwise following block is treated as a literal code block. Good to go otherwise, thanks!

```suggestion
  that a segment will be be buffered awaiting retransmission:
```

kartben reviewed (approved):
This is awesome to have - thanks @jukkar!
